velox/exec/RowContainer.h

```diff
@@ -1180,6 +1188,30 @@ class RowContainer {
   // Free any aggregates associated with the 'rows'.
   void freeAggregates(folly::Range<char**> rows);
 
+  // Free next row vectors associated with the 'rows'.
+  // If 'validateNextRowVector' is true, we must ensure that all duplicate rows
+  // are erased in a single call without any omissions. This check is
+  // unnecessary when clearing the container and initializing new rows.
+  void freeNextRowVectors(
+      folly::Range<char**> rows,
+      bool validateNextRowVector) {
+    if (!nextOffset_ || !hasDuplicateRows_) {
+      return;
+    }
+    for (auto row : rows) {
+      auto& vector = getNextRowVector(row);
+      if (vector) {
+        if (validateNextRowVector) {
+          for (auto next : *vector) {
+            VELOX_CHECK(rows.contains(next));
```

Comment:
Can we avoid O(N²) here?

Reply:
This function is used by some application directly so we need to do this check

Reply:
looks it's little hard to reduce the time complexity  with following the original approach.
I update the data struct to support row erasing. If value in nextOffset_ slot is not null, then it will be removed from corresponding next-row-vector and set it's nextOffset_ slot as null.
New data struct looks as below:
![image](https://github.com/facebookincubator/velox/assets/10524738/5036d4c5-3879-450d-9042-bd412238c201)